src/markov/mc_tools.jl

```diff
@@ -291,189 +291,156 @@ todense(::Type, A::Array) = A
-    k = length(init)*num_reps
-    X = Array(Int, ts_length, k)
-    X[1, :] = repmat(init, num_reps)
-
-    simulate!(mc, X)
-    return X
-end
+### Returns
 
+- `X::Vector` : Vector containing the sample path, with length 
+ts_length
 """
-Simulate time series of state transitions of the Markov chain `mc`.
-
-##### Arguments
-
-- `mc::MarkovChain` : MarkovChain instance.
-- `ts_length::Int` : Length of each simulation.
-- `init::Int` : Initial state.
-- `;num_reps::Int(1)` : Number of repetitions of simulation.
-
-##### Returns
 
-- `X::Matrix{Int}` : Array containing the sample paths as columns, of shape
-(ts_length, k), where k = num_reps.
+function simulate(mc::MarkovChain, ts_length::Int;
+                  init::Int=rand(1:n_states(mc)))
+    X = Array(eltype(mc.state_values), ts_length, 1)
+    simulate!(X, mc; init=init)
+    return vec(X)
+end
 
-"""
-simulate(mc::MarkovChain, ts_length::Int, init::Int; num_reps::Int=1) =
-    simulate(mc, ts_length, [init], num_reps=num_reps)
 
 """
-Simulate time series of state transitions of the Markov chain `mc`.
+Fill `X` with sample paths of the Markov chain `mc` as columns.
+The resulting matrix has the state values of `mc` as elements.
 
-##### Arguments
+### Arguments
 
 - `mc::MarkovChain` : MarkovChain instance.
-- `ts_length::Int` : Length of each simulation.
-- `;num_reps::Union{Int, Void}(nothing)` : Number of repetitions of simulation.
-
-##### Returns
-
-- `X::Matrix{Int}` : Array containing the sample paths as columns, of shape
-(ts_length, k), where k = num_reps.
-
+- `X::Matrix` : Preallocated matrix to be filled with sample paths
+of the Markov chain `mc`. The element types in `X` should be the 
+same as the type of the state values of `mc`
+- `;init=rand(1:n_states(mc))` : Can be one of the following
+    - blank: random initial condition for each chain
+    - scalar: same initial condition for each chain
+    - vector: cycle through the elements, applying each as an
+      initial condition until all columns have an initial condition 
+      (allows for more columns than initial conditions)
 """
-function simulate(mc::MarkovChain, ts_length::Int; num_reps::Int=1)
-    init = rand(1:size(mc.p, 1), num_reps)
-    return simulate(mc, ts_length, init)
-end
 
-"""
-Fill `X` with sample paths of the Markov chain `mc` as columns.
+function simulate!(X::Union{AbstractVector,AbstractMatrix},
+                   mc::MarkovChain; init=rand(1:n_states(mc), size(X, 2)))
 
-##### Arguments
+    # Note: eltype(X) must be the same as eltype(mc.state_values)
+    if eltype(X) != eltype(mc.state_values)
+        throw(ArgumentError("Types in X must be the same as the types of the
+        Markov state values"))
+    end
 
-- `mc::MarkovChain` : MarkovChain instance.
-- `X::Matrix{Int}` : Preallocated matrix of integers to be filled with sample
-paths of the markov chain `mc`. The elements in `X[1, :]` will be used as the
-initial states.
+    ts_length, k = size(X)
```

Comment:
This line will not actually work when x is an `AbstractVector`.

We should split it into

```julia
ts_length = size(X, 1)
k = size(X, 2)
```